### PR TITLE
Adding default event handler

### PR DESCRIPTION
**Pull request by tonymort**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/darts/event/service/impl/HandlerNotFoundException.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package uk.gov.hmcts.darts.event.service.impl;
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/event/service/impl/HandlerNotFoundException.java:1]: Should this be moved under `package uk.gov.hmcts.darts.event.exception`?

Comment (tonymort) [reply src/main/java/uk/gov/hmcts/darts/event/service/impl/HandlerNotFoundException.java:1]: yes and also should be using Chris B new exceptions. will update.

File: src/main/java/uk/gov/hmcts/darts/event/EventTypeRepository.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package uk.gov.hmcts.darts.event;
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/event/EventTypeRepository.java:1]: Should this be under `package uk.gov.hmcts.darts.common.repository` to be close to `EventEntity`?

File: src/main/java/uk/gov/hmcts/darts/event/controller/EventsController.java [hunk 3]
```diff
@@ -50,9 +50,8 @@ public class EventsController implements EventApi {
     public ResponseEntity<AddDocumentResponse> eventsPost(
         @Parameter(name = "DartsEvent", description = "") @Valid @RequestBody(required = false) DartsEvent dartsEvent
     ) {
-
         // send to darts-gateway for DAR PC Notify
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/event/controller/EventsController.java:53]: Please could you remove this redundant comment for me - since you have covered the complexities of when to handle the darNotify now. Thanks

File: src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.HttpStatus;
+import uk.gov.hmcts.darts.common.exception.DartsApiError;
+
+@Getter
+@RequiredArgsConstructor
+public enum EventError implements DartsApiError {
+
+    EVENT_DATA_NOT_FOUND(
+          "422",
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java:13]: Just wondering what made you choose this `errorTypeNumeric` over say "200"? :)

File: src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+          HttpStatus.BAD_REQUEST,
+          "Data on the event could be reconciled with Darts records");
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java:15]: Should this `title` say **not** reconciled? Or maybe starts with "Failed to reconcile..."?

File: src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.http.HttpStatus;
+import uk.gov.hmcts.darts.common.exception.DartsApiError;
+
+@Getter
+@RequiredArgsConstructor
+public enum EventError implements DartsApiError {
+
+    EVENT_DATA_NOT_FOUND(
+          "422",
+          HttpStatus.BAD_REQUEST,
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java:14]: I think this should be a 404 - NOT_FOUND status code?

File: src/main/java/uk/gov/hmcts/darts/event/service/impl/DefaultEventHandler.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    public void handle(final DartsEvent dartsEvent) {
+        final var actualCourtRoomEntity = courtroomRepository
+            .findByCourthouseNameAndCourtroomName(dartsEvent.getCourthouse(), dartsEvent.getCourtroom())
+            .orElseThrow(() -> new DartsApiException(
+                  EventError.EVENT_DATA_NOT_FOUND,
+                  format(NOT_FOUND_MESSAGE, dartsEvent.getCourthouse(), dartsEvent.getCourtroom())));
+
+        final var caseNumbers = dartsEvent.getCaseNumbers();
+        if (caseNumbers.size() > 1) {
+            log.info(format(MULTIPLE_CASE_NUMBERS, dartsEvent.getEventId(), join(", ", caseNumbers)));
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/event/service/impl/DefaultEventHandler.java:52]: Should this log message be a `WARN`?

File: src/main/java/uk/gov/hmcts/darts/common/entity/CaseEntity.java [hunk 8]
```diff
@@ -74,4 +78,11 @@ public class CaseEntity {
 
     @OneToMany(mappedBy = "courtCase")
     private List<AnnotationEntity> annotations;
+
+    @OneToMany(mappedBy = "courtCase")
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/common/entity/CaseEntity.java:82]: Do you think this relationship should have `cascade = {CascadeType.PERSIST, CascadeType.MERGE}`?

File: src/main/java/uk/gov/hmcts/darts/event/service/impl/DefaultEventHandler.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        var eventEntity = eventEntityFrom(dartsEvent);
+        eventEntity.setCourtroom(actualCourtRoomEntity);
+        eventRepository.save(eventEntity);
+
+        if (eitherTheHearingIsNewOrTheCourtroomIsDifferent(actualCourtRoomEntity, actualHearingEntity)) {
+            actualHearingEntity.setCourtroom(actualCourtRoomEntity);
+            darNotifyService.darNotify(dartsEvent, CASE_UPDATE);
+        }
+
+        caseRepository.save(actualCourtCaseEntity);
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/event/service/impl/DefaultEventHandler.java:77]: Why do we need to call `save` on `Case` - did we change something?

File: src/main/resources/db/migration/V1_16__add_court_house_to_case_table.sql [hunk 10]
```diff
@@ -0,0 +1,6 @@
+alter table darts.moj_case add moj_cth_id integer;
```
Comment (richardbull) [inline src/main/resources/db/migration/V1_16__add_court_house_to_case_table.sql:1]: These have the schema prefix `darts` which might cause a problem?
We haven't been using that so far, but maybe it's ok to use now.
I think this should just be pushed up into V1_13 to correct the schema, as we've went back to touch that script anyway and we seem to be blowing away our Docker volumes at the moment as these scripts may be short lived, if we go with a full schema baseline instead in S6+

File: src/main/java/uk/gov/hmcts/darts/event/service/impl/DefaultEventHandler.java (hunk 9 above)
Comment (tonymort) [reply src/main/java/uk/gov/hmcts/darts/event/service/impl/DefaultEventHandler.java:77]: Because we might have created if it wasn't in the database.  Its hard to follow I know.

File: src/integrationTest/resources/db/migration/V1_9999__integrationTestData.sql
Comment (richardbull) [inline src/integrationTest/resources/db/migration/V1_9999__integrationTestData.sql:1]: It would be better if we listed the columns on `INSERT INTO moj_case`

File: src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+          "Data on the event could not be reconciled with Darts records"),
+
+    EVENT_HANDLER_NOT_FOUND("422",
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java:17]: Please check this with @chrisbellingham-hmcts  - I'm pretty sure he means for these to be a distinct numeric within an error type, if you look at his example `AuthenticationError`. See what he says about the status code too 👍

File: src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+          HttpStatus.BAD_REQUEST,
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java:18]: Maybe you want `UNPROCESSABLE_ENTITY` in both places if it matters

File: src/main/java/uk/gov/hmcts/darts/event/service/impl/EventHandlerBase.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import uk.gov.hmcts.darts.common.entity.EventEntity;
+import uk.gov.hmcts.darts.common.entity.EventTypeEntity;
+import uk.gov.hmcts.darts.common.repository.EventTypeRepository;
+import uk.gov.hmcts.darts.event.model.DartsEvent;
+import uk.gov.hmcts.darts.event.service.EventHandler;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.PostConstruct;
```
Comment (richardbull) [inline src/main/java/uk/gov/hmcts/darts/event/service/impl/EventHandlerBase.java:13]: I think this should be `jakarta` package if we need to use `PostConstruct`?

File: src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java (hunk 11 above)
Comment (tonymort) [reply src/main/java/uk/gov/hmcts/darts/event/exception/EventError.java:17]: Yeah think you must be right about the errorTypeNumeric

**Review by richardbull — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/darts/event/service/impl/EventHandlerBase.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (tonymort) [reply src/main/java/uk/gov/hmcts/darts/event/service/impl/EventHandlerBase.java:13]: checked this. javax is fine with spring.
